gnoi/os/server.go

```diff
@@ -62,3 +73,106 @@ func (s *Server) Verify(ctx context.Context, _ *pb.VerifyRequest) (*pb.VerifyRes
 		ActivationFailMessage: s.manager.activationFailMessage,
 	}, nil
 }
+
+// Install receives an OS package, validates the package and then installs the package.
+func (s *Server) Install(stream pb.OS_InstallServer) error {
+	var request *pb.InstallRequest
+	var response *pb.InstallResponse
+	var err error
+	if request, err = stream.Recv(); err != nil {
+		return err
+	}
+	utils.LogProto(request)
+	transferRequest := request.GetTransferRequest()
+	if transferRequest == nil {
+		return errors.New("Failed to receive TransferRequest")
+	}
+	if s.manager.IsRunning(transferRequest.Version) {
+		response = &pb.InstallResponse{Response: &pb.InstallResponse_InstallError{
+			InstallError: &pb.InstallError{Type: pb.InstallError_INSTALL_RUN_PACKAGE},
+		}}
+		utils.LogProto(response)
+		if err = stream.Send(response); err != nil {
+			return err
+		}
+		return errors.New("Attempting to force transfer an OS of the same version as the currently running OS")
+	}
+	if version := transferRequest.Version; s.manager.IsInstalled(version) {
+		response = &pb.InstallResponse{Response: &pb.InstallResponse_Validated{
+			Validated: &pb.Validated{
+				Version: version,
+			},
+		}}
+		utils.LogProto(response)
+		if err = stream.Send(response); err != nil {
+			return err
+		}
+		return nil
+	}
+	select {
+	case <-s.installToken:
+		defer func() {
+			s.installToken <- true
+		}()
+	default:
+		response = &pb.InstallResponse{Response: &pb.InstallResponse_InstallError{InstallError: &pb.InstallError{Type: pb.InstallError_INSTALL_IN_PROGRESS}}}
+		utils.LogProto(response)
+		if err = stream.Send(response); err != nil {
+			return err
+		}
+		return errors.New("Another install is already in progress")
+	}
+
+	response = &pb.InstallResponse{Response: &pb.InstallResponse_TransferReady{TransferReady: &pb.TransferReady{}}}
+	utils.LogProto(response)
+	if err = stream.Send(response); err != nil {
+		return err
+	}
+
+	bb, err := ReceiveOS(stream)
+	if err != nil {
+		return err
+	}
+	mockOS, err, errResponse := mockos.ValidateOS(bb)
+	if err != nil {
+		response = &pb.InstallResponse{Response: errResponse}
+		utils.LogProto(response)
+		if err = stream.Send(response); err != nil {
+			return err
+		}
+		return err
+	}
+	s.manager.Install(mockOS.Version, mockOS.ActivationFailMessage)
+	return nil
+}
+
+// ReceiveOS receives and parses requests from stream, storing OS package into a buffer, and updating the progress.
+func ReceiveOS(stream pb.OS_InstallServer) (*bytes.Buffer, error) {
+	bb := new(bytes.Buffer)
+	prev := 0
+	for {
+		in, err := stream.Recv()
+		if err != nil {
+			return nil, err
+		}
+		utils.LogProto(in)
```

Comment:
Only log the proto when it is not a data transfer else we flood screen/log =)

Reply:
I think you removed the wrong log! =)
The transfer progress is fine. The receiving bytes from the client can go.

Reply:
Definitely, my bad, sorry!